<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Dissecting tumor transcriptional heterogeneity from multi-tumor single-cell RNA-seq data using GBCD • gbcd</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Dissecting tumor transcriptional heterogeneity from multi-tumor single-cell RNA-seq data using GBCD">
<meta property="og:description" content="gbcd">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">gbcd</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.1-5</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">Home</a>
</li>
<li>
  <a href="../articles/index.html">Vignettes</a>
</li>
<li>
  <a href="../reference/index.html">Functions</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/stephenslab/gbcd" class="external-link">Source</a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Dissecting tumor transcriptional heterogeneity
from multi-tumor single-cell RNA-seq data using GBCD</h1>
                        <h4 data-toc-skip class="author">Yusha Liu</h4>
            
            <h4 data-toc-skip class="date">2024-05-08</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/stephenslab/gbcd/blob/HEAD/vignettes/hnscc.Rmd" class="external-link"><code>vignettes/hnscc.Rmd</code></a></small>
      <div class="hidden name"><code>hnscc.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="overview">Overview<a class="anchor" aria-label="anchor" href="#overview"></a>
</h2>
<p>This vignette shows how to apply “generalized binary covariance
decomposition” (GBCD) to jointly analyze single-cell RNA-seq (scRNA-seq)
data from <em>malignant cells</em> collected from multiple patients
and/or studies, using the <a href="https://github.com/stephenslab/gbcd" class="external-link">gbcd</a> package.</p>
<p>GBCD can effectively dissect tumor transcriptional heterogeneity into
patient/study-specific and shared gene expression programs (GEPs). GBCD
is “unsupervised” in that, unlike tumor-by-tumor (e.g., <a href="https://www.sciencedirect.com/science/article/pii/S0092867417312709" class="external-link">Puram
et al. (2017)</a>) and many harmonization approaches (e.g., <a href="https://portals.broadinstitute.org/harmony" class="external-link">Harmony</a>, <a href="https://github.com/welch-lab/liger" class="external-link">Liger</a>), it does not use
information about which cell comes from which tumor or study. Instead,
GBCD only requires the combined scRNA-seq data from all tumors, which
are stored as an <span class="math inline">\(N \times J\)</span> matrix
<span class="math inline">\(Y\)</span> of expression values with entries
<span class="math inline">\(y_{ij}\)</span>, where <span class="math inline">\(i=1,\dots,N\)</span> indexes malignant cells and
<span class="math inline">\(j=1,\dots,J\)</span> indexes genes. In
typical applications, <span class="math inline">\(Y\)</span> contains
log-transformed pseudo-count-modified UMI counts (“log-pc counts”).</p>
<p>GBCD ultimately yields a decomposition of the expression data matrix
<span class="math inline">\(Y\)</span> into matrices <span class="math inline">\(L\)</span> and <span class="math inline">\(F\)</span> such that <span class="math inline">\(Y
\approx L F^T\)</span>, or equivalently, <span class="math display">\[y_{ij} \approx \sum_{k=1}^K l_{ik}
f_{jk}.\]</span> The <span class="math inline">\(K\)</span> components
should be interpretable as GEPs, with <span class="math inline">\(l_{ik}\)</span> representing the membership of
cell <span class="math inline">\(i\)</span> in GEP <span class="math inline">\(k\)</span>, and <span class="math inline">\(f_{jk}\)</span> representing the effect of GEP
<span class="math inline">\(k\)</span> on the expression of gene <span class="math inline">\(j\)</span>. When <span class="math inline">\(y_{ij}\)</span> are log-pc counts, each <span class="math inline">\(f_{jk}\)</span> approximately represents the
log-fold change (LFC) associated with membership in GEP <span class="math inline">\(k\)</span>, so we refer to the <span class="math inline">\(f_{jk}\)</span> values as LFCs, and to the vector
of LFCs <span class="math inline">\((f_{1k}, \dots, f_{Jk})^T\)</span>
as the “signature” of GEP <span class="math inline">\(k\)</span>.</p>
<p>We begin our analysis by loading the needed R packages.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://wilkelab.org/cowplot/" class="external-link">cowplot</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">RColorBrewer</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggrepel.slowkow.com/" class="external-link">ggrepel</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">pheatmap</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">gridExtra</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/willwerscheid/flashier" class="external-link">flashier</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/stephenslab/gbcd" class="external-link">gbcd</a></span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="the-hnscc-dataset">The HNSCC dataset<a class="anchor" aria-label="anchor" href="#the-hnscc-dataset"></a>
</h2>
<p>This dataset contained gene expression data for <span class="math inline">\(n=2,176\)</span> malignant cells collected from
primary tumors from 10 HNSCC patients and matching lymph node (LN)
metastases from 5 of these patients. <a href="https://www.sciencedirect.com/science/article/pii/S0092867417312709" class="external-link">Puram
et al. (2017)</a> found that each of these 10 patients clearly mapped to
a molecular subtype of HNSCC, whose signatures were previously defined
by analysis of bulk expression data of 279 <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4311405/" class="external-link">TCGA</a>
HNSCC tumors.</p>
<p>Unlike more recently generated data sets, the HNSCC data were not UMI
counts; rather, they were read counts produced by SMART-Seq2. Following
<a href="https://www.sciencedirect.com/science/article/pii/S0092867417312709" class="external-link">Puram
et al. (2017)</a>, we defined the transformed counts as <span class="math inline">\(y_{ij} = \log_2(1 +
\mathrm{TPM}_{ij}/10)\)</span>, where <span class="math inline">\(\mathrm{TPM}_{ij}\)</span> was the
transcript-per-million (TPM) value for gene <span class="math inline">\(j\)</span> in cell <span class="math inline">\(i\)</span>.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">hnscc</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">Y</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">info</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="visualize-hnscc-data-using-t-sne">Visualize HNSCC data using t-SNE<a class="anchor" aria-label="anchor" href="#visualize-hnscc-data-using-t-sne"></a>
</h2>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">hnscc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SeuratObject/man/CreateSeuratObject.html" class="external-link">CreateSeuratObject</a></span><span class="op">(</span>counts <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="va">Y</span><span class="op">)</span>, project <span class="op">=</span> <span class="st">"HNSCC"</span>, meta.data <span class="op">=</span> <span class="va">info</span><span class="op">)</span></span>
<span><span class="va">hnscc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindVariableFeatures.html" class="external-link">FindVariableFeatures</a></span><span class="op">(</span><span class="va">hnscc</span>, selection.method <span class="op">=</span> <span class="st">"vst"</span>, nfeatures <span class="op">=</span> <span class="fl">5000</span><span class="op">)</span></span>
<span><span class="va">all.genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">hnscc</span><span class="op">)</span></span>
<span><span class="va">hnscc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/ScaleData.html" class="external-link">ScaleData</a></span><span class="op">(</span><span class="va">hnscc</span>, features <span class="op">=</span> <span class="va">all.genes</span><span class="op">)</span></span>
<span><span class="va">hnscc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/RunPCA.html" class="external-link">RunPCA</a></span><span class="op">(</span><span class="va">hnscc</span>, features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/SeuratObject/man/VariableFeatures.html" class="external-link">VariableFeatures</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">hnscc</span><span class="op">)</span>, npcs <span class="op">=</span> <span class="fl">50</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="va">hnscc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/RunTSNE.html" class="external-link">RunTSNE</a></span><span class="op">(</span><span class="va">hnscc</span>, dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">50</span><span class="op">)</span></span></code></pre></div>
<p>We plot the 2-D <em>t</em>-SNE embedding of these <span class="math inline">\(n=2,176\)</span> malignant cells, which are
colored by patient of origin and tumor stage (primary tumor, LN
metastasis), and shaped by tumor molecular subtype. Due to the presence
of strong inter-tumor heterogeneity, these cells demonstrate strong
patient effects that are typical of cancer data; the major structure in
the <em>t</em>-SNE visualization is the clustering of the cells by
patient.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">hnscc</span>, label <span class="op">=</span> <span class="cn">TRUE</span>, repel <span class="op">=</span> <span class="cn">TRUE</span>, pt.size <span class="op">=</span> <span class="fl">1</span>, reduction <span class="op">=</span> <span class="st">"tsne"</span>, group.by <span class="op">=</span> <span class="st">"subject"</span>, shape.by <span class="op">=</span> <span class="st">"subtype"</span>, cols <span class="op">=</span> <span class="va">subject_col</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guides.html" class="external-link">guides</a></span><span class="op">(</span>shape <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guide_legend.html" class="external-link">guide_legend</a></span><span class="op">(</span>override.aes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">3</span><span class="op">)</span><span class="op">)</span>, ncol <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>text <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">12</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">""</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_shape_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">15</span>, <span class="fl">16</span>, <span class="fl">18</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="apply-gbcd-to-estimate-gep-memberships-and-signatures">Apply GBCD to estimate GEP memberships and signatures<a class="anchor" aria-label="anchor" href="#apply-gbcd-to-estimate-gep-memberships-and-signatures"></a>
</h2>
<p>To fit GBCD, users need to specify <span class="math inline">\(Kmax\)</span>, an upper bound of the number of
components <span class="math inline">\(K\)</span>, just for
initialization purposes; the final number of GEPs returned by GBCD is
close to but often not equal to (and can be larger than) <span class="math inline">\(Kmax\)</span>. Generally speaking, a larger <span class="math inline">\(Kmax\)</span> allows for identifying finer
structure in tumor transcriptional heterogeneity, at the expense of
higher computational cost. A good choice of <span class="math inline">\(Kmax\)</span> depends on the complexity of the
analyzed data (e.g., the number of patients and studies contained), and
is often unknown in advance. Based on our experience, a value of <span class="math inline">\(Kmax\)</span> between 10 and 40 usually works
well, and users are encouraged to explore and compare the results for
different values of <span class="math inline">\(Kmax\)</span>.</p>
<p>Users also need to specify a prior for GEP membership values. GBCD
assigns a “generalized binary” (GB) prior independently to each entry of
<span class="math inline">\(L\)</span>, <span class="math display">\[l_{ik}\sim (1-\pi_k^l) \delta_0 + \pi_k^l \:
N_{+} (\mu_k,
\sigma_k^2),\]</span> where the ratio <span class="math inline">\(\sigma_k/\mu_k\)</span> is fixed at some
pre-specified small value. In this application we set <span class="math inline">\(\sigma_k/\mu_k = 0.04\)</span> for each <span class="math inline">\(k\)</span>. A smaller ratio is generally
recommended for larger datasets containing more cells, and a value
between 0.01 to 0.1 usually works well.</p>
<p>For time reason, in the following call, we set
<code>Kmax = 12</code>, and set the number of backfit iterations
respectively to <code>maxiter1 = 100</code> (for initializing <span class="math inline">\(L\)</span>), <code>maxiter2 = 50</code> (for
estimating <span class="math inline">\(L\)</span>), and
<code>maxiter3 = 100</code> (for estimating <span class="math inline">\(F\)</span>). However, users are encouraged to
explore larger values of <code>Kmax</code> and set
<code>maxiter1</code>, <code>maxiter2</code>, <code>maxiter3</code> to
their default values, which take longer to run.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res.gbcd</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/fit_gbcd.html">fit_gbcd</a></span><span class="op">(</span>Y <span class="op">=</span> <span class="va">Y</span>, Kmax <span class="op">=</span> <span class="fl">12</span>, maxiter1 <span class="op">=</span> <span class="fl">100</span>, maxiter2 <span class="op">=</span> <span class="fl">50</span>, maxiter3 <span class="op">=</span> <span class="fl">100</span>,</span>
<span>                     prior <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/flashier/man/flash_ebnm.html" class="external-link">flash_ebnm</a></span><span class="op">(</span>prior_family <span class="op">=</span> <span class="st">"generalized_binary"</span>, scale <span class="op">=</span> <span class="fl">0.04</span><span class="op">)</span>, verbose <span class="op">=</span> <span class="fl">0</span><span class="op">)</span></span></code></pre></div>
<p>Note that users can also specify other nonnegative priors for GEP
membership values, as long as they are defined in the <a href="https://stephenslab.github.io/ebnm/" class="external-link">ebnm</a> package. For
example, users can specify a point exponential prior by setting
<code>prior = ebnm::ebnm_point_exponential</code> in the call to
<code>fit_gbcd</code>. However, we find that the GB prior usually
performs the best empirically to dissect tumor transcriptional
heterogeneity from multiple samples and/or studies.</p>
<p>The above call takes about 10 minutes to run on a laptop.</p>
</div>
<div class="section level2">
<h2 id="interpret-gbcd-results-from-the-hnscc-data">Interpret GBCD results from the HNSCC data<a class="anchor" aria-label="anchor" href="#interpret-gbcd-results-from-the-hnscc-data"></a>
</h2>
<p>Take a look at a summary of the GBCD results.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">res.gbcd</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="gep-membership-estimates">GEP membership estimates<a class="anchor" aria-label="anchor" href="#gep-membership-estimates"></a>
</h2>
<p>We now look at the matrix of GEP membership estimates <span class="math inline">\(L\)</span> with entries <span class="math inline">\(l_{ik}\)</span> representing the membership of
cell <span class="math inline">\(i\)</span> in GEP <span class="math inline">\(k\)</span>. Here we show the top rows of <span class="math inline">\(L\)</span>.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">round</a></span><span class="op">(</span><span class="va">res.gbcd</span><span class="op">$</span><span class="va">L</span>, <span class="fl">3</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Create a heatmap to visualize GEP membership estimates.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">### add the sample and molecular subtype annotation of cells to visualize the GEP memberships</span></span>
<span><span class="va">anno</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>subject<span class="op">=</span><span class="va">info</span><span class="op">$</span><span class="va">subject</span>, subtype<span class="op">=</span><span class="va">info</span><span class="op">$</span><span class="va">subtype</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">anno</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">res.gbcd</span><span class="op">$</span><span class="va">L</span><span class="op">)</span></span>
<span><span class="va">anno_colors</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>subject<span class="op">=</span><span class="va">subject_col</span>, subtype<span class="op">=</span><span class="va">subtype_col</span><span class="op">)</span></span>
<span><span class="va">cols</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/grDevices/colorRamp.html" class="external-link">colorRampPalette</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"gray96"</span>, <span class="st">"red"</span><span class="op">)</span><span class="op">)</span><span class="op">(</span><span class="fl">50</span><span class="op">)</span></span>
<span><span class="va">brks</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1</span>, <span class="fl">0.02</span><span class="op">)</span></span>
<span></span>
<span><span class="co">### plot the annotated heatmap of GEP memberships</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/pheatmap/man/pheatmap.html" class="external-link">pheatmap</a></span><span class="op">(</span><span class="va">res.gbcd</span><span class="op">$</span><span class="va">L</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/order.html" class="external-link">order</a></span><span class="op">(</span><span class="va">anno</span><span class="op">$</span><span class="va">subject</span><span class="op">)</span>, <span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span><span class="op">]</span>, cluster_rows <span class="op">=</span> <span class="cn">FALSE</span>, cluster_cols <span class="op">=</span> <span class="cn">FALSE</span>, show_rownames <span class="op">=</span> <span class="cn">FALSE</span>, annotation_row <span class="op">=</span> <span class="va">anno</span>, annotation_colors <span class="op">=</span> <span class="va">anno_colors</span>, annotation_names_row <span class="op">=</span> <span class="cn">FALSE</span>, angle_col <span class="op">=</span> <span class="fl">45</span>, fontsize <span class="op">=</span> <span class="fl">10</span>, color <span class="op">=</span> <span class="va">cols</span>, breaks <span class="op">=</span> <span class="va">brks</span>, main <span class="op">=</span> <span class="st">""</span><span class="op">)</span></span></code></pre></div>
<p>This heatmap shows the membership values of the <span class="math inline">\(n=2,176\)</span> cells (rows) for 20 GEPs
(columns; intercept excluded). Cells are arranged top-to-bottom by tumor
molecular subtype and patient of origin. Membership values are rescaled
separately for GEPs so that the maximum membership for each GEP is
always 1.</p>
<p>From this heatmap, we see that GEPs 1-3 correspond closely to the
molecular subtypes previously defined based on bulk RNA-seq data: GEP1
is largely active only in cells from the 2 classical patients, GEP2 is
mainly active in cells from the 7 basal patients, and GEP3 is mainly
active in cells from the 1 atypical patient. This demonstrates that GBCD
can extract the molecular subtype information <em>de novo</em> from the
single cell data alone.</p>
<p>Among other GEPs, some are active predominantly in an individual
patient and can thus be interpreted as patient-specific GEPs, such as
GEPs 8,9,11; other GEPs are shared across patients but do not correspond
to molecular subtypes. The degree of patient-specific (vs. shared across
patients) expression for each GEP can be quantified using the function
<code>gep_annotation</code>.</p>
</div>
<div class="section level2">
<h2 id="gep-signature-estimates">GEP signature estimates<a class="anchor" aria-label="anchor" href="#gep-signature-estimates"></a>
</h2>
<p>We next look at the matrix of GEP signature estimates <span class="math inline">\(F\)</span> with each entry <span class="math inline">\(f_{jk}\)</span> approximately representing the
<span class="math inline">\(\text{log}_2\)</span>-fold change associated
with membership in GEP <span class="math inline">\(k\)</span>.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">round</a></span><span class="op">(</span><span class="va">res.gbcd</span><span class="op">$</span><span class="cn">F</span><span class="op">$</span><span class="va">lfc</span>, <span class="fl">3</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Create a volcano plot to visualize the gene signature for a given GEP
(here we plot GEP2 as an example), which also makes use of measures of
significance of LFCs produced by GBCD, including <em>z-score</em> and
<em>local false sign rate</em>, or <a href="https://academic.oup.com/biostatistics/article/18/2/275/2557030" class="external-link"><em>lfsr</em></a>.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pdat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>gene <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">res.gbcd</span><span class="op">$</span><span class="cn">F</span><span class="op">$</span><span class="va">lfc</span><span class="op">)</span>, </span>
<span>                   lfc <span class="op">=</span> <span class="va">res.gbcd</span><span class="op">$</span><span class="cn">F</span><span class="op">$</span><span class="va">lfc</span><span class="op">[</span>, <span class="st">"GEP2"</span><span class="op">]</span>, </span>
<span>                   z <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">abs</a></span><span class="op">(</span><span class="va">res.gbcd</span><span class="op">$</span><span class="cn">F</span><span class="op">$</span><span class="va">z_score</span><span class="op">[</span>, <span class="st">"GEP2"</span><span class="op">]</span><span class="op">)</span>, </span>
<span>                   lfsr <span class="op">=</span> <span class="va">res.gbcd</span><span class="op">$</span><span class="cn">F</span><span class="op">$</span><span class="va">lfsr</span><span class="op">[</span>, <span class="st">"GEP2"</span><span class="op">]</span>,</span>
<span>                   stringsAsFactors <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="va">pdat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/transform.html" class="external-link">transform</a></span><span class="op">(</span><span class="va">pdat</span>, lfsr <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/cut.html" class="external-link">cut</a></span><span class="op">(</span><span class="va">lfsr</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">1</span>, <span class="fl">0.001</span>, <span class="fl">0.01</span>, <span class="fl">0.05</span>, <span class="cn">Inf</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">rows</span>  <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/with.html" class="external-link">with</a></span><span class="op">(</span><span class="va">pdat</span>, <span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="op">!</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">abs</a></span><span class="op">(</span><span class="va">lfc</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fu"><a href="https://rdrr.io/r/stats/quantile.html" class="external-link">quantile</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">abs</a></span><span class="op">(</span><span class="va">lfc</span><span class="op">)</span>, <span class="fl">0.998</span><span class="op">)</span> <span class="op">|</span> <span class="op">(</span><span class="va">z</span> <span class="op">&gt;</span> <span class="fl">10</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">pdat</span><span class="op">[</span><span class="va">rows</span>, <span class="st">"gene"</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">""</span></span>
<span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">pdat</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">lfc</span>, y <span class="op">=</span> <span class="va">z</span>, color <span class="op">=</span> <span class="va">lfsr</span>, label <span class="op">=</span> <span class="va">gene</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggrepel.slowkow.com/reference/geom_text_repel.html" class="external-link">geom_text_repel</a></span><span class="op">(</span>color <span class="op">=</span> <span class="st">"black"</span>, size <span class="op">=</span> <span class="fl">3</span>, segment.color <span class="op">=</span> <span class="st">"black"</span>, segment.size <span class="op">=</span> <span class="fl">0.25</span>, min.segment.length <span class="op">=</span> <span class="fl">0</span>, max.overlaps <span class="op">=</span> <span class="cn">Inf</span>, na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_color_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"coral"</span>, <span class="st">"orange"</span>, <span class="st">"gold"</span>, <span class="st">"deepskyblue"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>x <span class="op">=</span> <span class="st">"log-fold change"</span>, y <span class="op">=</span> <span class="st">"|posterior z-score|"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guides.html" class="external-link">guides</a></span><span class="op">(</span>colour <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guide_legend.html" class="external-link">guide_legend</a></span><span class="op">(</span>override.aes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">4</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>plot.title <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>hjust <span class="op">=</span> <span class="fl">0.5</span>, size <span class="op">=</span> <span class="fl">24</span><span class="op">)</span>, axis.text <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">20</span><span class="op">)</span>, axis.title <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">20</span><span class="op">)</span>, legend.title <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">20</span><span class="op">)</span>, legend.text <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">16</span><span class="op">)</span>, legend.position <span class="op">=</span> <span class="st">"bottom"</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"Volcano plot of gene signature for GEP2"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="session-info">Session info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<p>This was the version of R and the packages that were used to generate
these results.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Yusha Liu, Peter Carbonetto, Jason Willwerscheid, Matthew Stephens.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
