<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Single-cell RNA-seq from HNSCC Tumor Cells — hnscc • gbcd</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Single-cell RNA-seq from HNSCC Tumor Cells — hnscc"><meta property="og:description" content="This dataset contains transcriptome profiles of
malignant cells generated via single-cell RNA sequencing. The cells
were collected from primary tumors in 10 HNSCC patients and
matching lymph node (LN) metastases from 5 of these patients.
Puram et al (2017) found that each of these 10 patients
clearly mapped to a molecular subtype of HNSCC, whose signatures
were previously defined by analysis of bulk expression data of 279
TCGA HNSCC tumors.
The data are normalized, log-transformed counts for 17,113 genes in
2,176 cells: \(y_{ij} = \log_2(1 + TPM_{ij}/10)\), where
\(TPM_{ij}\) is the transcript-per-million (TPM) value for gene
\(j\) in cell \(i\). The counts are stored as a 2,176 x 17,113
sparse matrix.
These data are used in the vignette to illustrate how gbcd
can be used to analyze single-cell RNA-seq data derived from
multiple tumor samples."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">gbcd</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.2-02</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../index.html">Home</a>
</li>
<li>
  <a href="../articles/index.html">Vignettes</a>
</li>
<li>
  <a href="../reference/index.html">Functions</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/stephenslab/gbcd" class="external-link">Source</a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Single-cell RNA-seq from HNSCC Tumor Cells</h1>
    <small class="dont-index">Source: <a href="https://github.com/stephenslab/gbcd/blob/HEAD/R/hnscc.R" class="external-link"><code>R/hnscc.R</code></a></small>
    <div class="hidden name"><code>hnscc.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>This dataset contains transcriptome profiles of
malignant cells generated via single-cell RNA sequencing. The cells
were collected from primary tumors in 10 HNSCC patients and
matching lymph node (LN) metastases from 5 of these patients.
Puram <em>et al</em> (2017) found that each of these 10 patients
clearly mapped to a molecular subtype of HNSCC, whose signatures
were previously defined by analysis of bulk expression data of 279
TCGA HNSCC tumors.</p>
<p>The data are normalized, log-transformed counts for 17,113 genes in
2,176 cells: \(y_{ij} = \log_2(1 + TPM_{ij}/10)\), where
\(TPM_{ij}\) is the transcript-per-million (TPM) value for gene
\(j\) in cell \(i\). The counts are stored as a 2,176 x 17,113
sparse matrix.</p>
<p>These data are used in the vignette to illustrate how gbcd
can be used to analyze single-cell RNA-seq data derived from
multiple tumor samples.</p>
    </div>


    <div id="format">
    <h2>Format</h2>
    <p><code>hnscc</code> is a list with the following elements:</p>
<dl><dt>Y</dt>
<dd><p>2,176 x 17,113 sparse matrix of normalized,
  log-transformed counts, with rows corresponding to cells and
  columns corresponding to genes.</p></dd>


<dt>info</dt>
<dd><p>Data frame containing information about the cells,
  including patient identity, primary/metastatic status and
  molecular subtype of the tumor sample.</p></dd>


<dt>sample_col</dt>
<dd><p>Color used in plots to indicate the different
  tumor samples.</p></dd>


<dt>subtype_col</dt>
<dd><p>Color used in plots to indicate the different
  molecular subtypes.</p></dd>

</dl></div>
    <div id="references">
    <h2>References</h2>
    <p>S.V. Puram <em>et al</em> (2017). Single-cell transcriptomic analysis of 
  primary and metastatic tumor ecosystems in head and neck cancer
  <em>Cell</em> <b>171</b>, 1611-1624. <a href="https://doi.org/10.1016/j.cell.2017.10.044" class="external-link">doi:10.1016/j.cell.2017.10.044</a></p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://Matrix.R-forge.R-project.org" class="external-link">Matrix</a></span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">hnscc</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Yusha Liu, Peter Carbonetto, Jason Willwerscheid, Matthew Stephens.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

